### PR TITLE
Use typespecs when possible to document types

### PR DESCRIPTION
**Pull request by whatyouhide**

### CODE REVIEW

**Comment by josevalim** [comment]
I am actually not the biggest fan because I think it makes understanding the options more complex, specially if you are not familar with typespecs.

Maybe we could support `NimbleOptions.spec` macro though that returns the spec for a given type?

**Comment by whatyouhide** [comment]
@josevalim the macro is a good idea in and of its own, I'll work on that.

When it comes to the docs, I think this makes it easier because we can express more complex types, with type documentation linked to them, in a bit easier syntax. The non-typespec docs can be verbose. We can think about it for a bit, I'll work on `NimbleOptions.spec/1` for now 😉

**Comment by whatyouhide** [comment]
@josevalim I added the new `NimbleOptions.option_type_union/1` function. 👍 Ready for review!

File: lib/nimble_options.ex [hunk 1]
```diff
@@ -377,6 +391,65 @@ defmodule NimbleOptions do
...
+
+      @type option() :: unquote(NimbleOptions.option_type_union(schema))
+
+  The code above would essentially compile to:
+
+      @type option() :: {:int, integer()} | {:number, integer() | float()}
+
+  """
+  # TODO: remove when we depend on Elixir 1.7+
+  if Version.match?(System.version(), "~> 1.7") do
```
Comment (josevalim) [inline lib/nimble_options.ex:438]: We can totally depend on ~> 1.7 I would say. Even 1.7 is no longer supported?

File: lib/nimble_options.ex [hunk 2]
```diff
@@ -377,6 +391,65 @@ defmodule NimbleOptions do
...
+    @doc since: "0.5.0"
+  end
+
+  @spec option_type_union(schema() | t()) :: Macro.t()
+  def option_type_union(schema)
```
Comment (josevalim) [inline lib/nimble_options.ex:443]: I would call it `option_type`only. `union` feels like an implementation detail?

File: lib/nimble_options.ex [hunk 3]
```diff
@@ -47,6 +47,20 @@ defmodule NimbleOptions do
...
+          doc: """
+          The type doc to use *in the documentation* for the option item. If `false`,
+          no type documentation is added to the item. If it's a string, it can be
+          anything. For example, you can use `"a list of PIDs"`, or you can use
+          a typespec reference that ExDoc can link to the type definition, such as
+          `` "`t:binary/0`" ``. You can use Markdown in this documentation. If the
+          `:type_doc` option is not present, NimbleOptions tries to produce a type
+          documentation automatically if it can do it unambiguously. For example,
+          if `type: :integer`, NimbleOptions will use `t:integer/0` as the
+          auto-generated typespec.
```
Comment (josevalim) [inline lib/nimble_options.ex:61]:
```suggestion
          auto-generated type doc.
```

File: lib/nimble_options.ex (hunk 2 above)
Comment (josevalim) [reply lib/nimble_options.ex:443]: Or maybe call it `option_typespec` to make it clear it is about typespecs?

File: lib/nimble_options/docs.ex [hunk 4]
```diff
@@ -98,19 +85,37 @@ defmodule NimbleOptions.Docs do
...
+  defp get_raw_type_str(:atom), do: "`t:atom/0`"
+  defp get_raw_type_str(:integer), do: "`t:integer/0`"
+  defp get_raw_type_str(:non_neg_integer), do: "`t:non_neg_integer/0`"
+  defp get_raw_type_str(:pos_integer), do: "`t:pos_integer/0`"
+  defp get_raw_type_str(:float), do: "`t:float/0`"
+  defp get_raw_type_str(:string), do: "`t:binary/0`"
+  defp get_raw_type_str(:keyword_list), do: "`t:keyword/0`"
+  defp get_raw_type_str(:non_empty_keyword_list), do: "`t:keyword/0`"
+  defp get_raw_type_str({:keyword_list, _keys}), do: "`t:keyword/0`"
+  defp get_raw_type_str({:non_empty_keyword_list, _keys}), do: "`t:keyword/0`"
```
Comment (josevalim) [inline lib/nimble_options/docs.ex:115]:
```suggestion
  defp get_raw_type_str(:keyword_list), do: "`t:keyword/0`"
  defp get_raw_type_str(:non_empty_keyword_list), do: "non-empty `t:keyword/0`"
  defp get_raw_type_str({:keyword_list, _keys}), do: "`t:keyword/0`"
  defp get_raw_type_str({:non_empty_keyword_list, _keys}), do: "non-empty `t:keyword/0`"
```

File: lib/nimble_options/docs.ex [hunk 5]
```diff
@@ -98,19 +85,37 @@ defmodule NimbleOptions.Docs do
...
+  defp get_raw_type_str(:reference), do: "`t:reference/0`"
+  defp get_raw_type_str(:pid), do: "`t:pid/0`"
+  defp get_raw_type_str(:timeout), do: "`t:timeout/0`"
+  defp get_raw_type_str(:boolean), do: "`t:boolean/0`"
+  defp get_raw_type_str(:atom), do: "`t:atom/0`"
+  defp get_raw_type_str(:integer), do: "`t:integer/0`"
+  defp get_raw_type_str(:non_neg_integer), do: "`t:non_neg_integer/0`"
+  defp get_raw_type_str(:pos_integer), do: "`t:pos_integer/0`"
+  defp get_raw_type_str(:float), do: "`t:float/0`"
+  defp get_raw_type_str(:string), do: "`t:binary/0`"
```
Comment (josevalim) [inline lib/nimble_options/docs.ex:111]:
```suggestion
  defp get_raw_type_str(:string), do: "`t:String.t/0`"
```

File: lib/nimble_options.ex (hunk 2 above)
Comment (josevalim) [reply lib/nimble_options.ex:443]: Yeah, let's go with `option_typespec`.

Comment (whatyouhide) [reply lib/nimble_options.ex:443]: Sounds good. 👍

**Comment by whatyouhide** [comment]
@josevalim ready for another round 😉

**Review by josevalim — APPROVED** [review]

**Pull request merged**